_posts/2020-11-12-deconvolution.md

```diff
@@ -0,0 +1,435 @@
+---
+layout: post
+title: Image Analysis Redux
+tagline: Dask + CuPy + RL
+author: John Kirkham (NVIDIA) and Ben Zaitlen (NVIDIA)
+theme: twitter
+---
+{% include JB/setup %}
+
+Summary
+-------
+
+[Last year](https://blog.dask.org/2019/08/09/image-itk) we experimented with
+Dask/ITK/Scikit-Image to perform large scale image analysis on a stack of 3D
+images.  Specifically, we looked at deconvolution, a common method to _deblur_
+images.  Now, a year later, we return to these experiments with a better
+understanding of how Dask and CuPy can interact, enhanced serialization
+methods, and support from the open-source community. This post looks at the
+following:
+
+1. Implementing a common deconvolution method for CPU + GPU
+2. Leveraging Dask to perform deconvolution on a larger dataset
+3. Exploring the results with the Napari image viewer
+
+Image Analysis Redux
+--------------------
+
+Previously we used the [Richardson Lucy
+(RL)](https://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution)
+deconvolution algorithm from ITK and
+[Scikit-Image](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/deconvolution.py#L329).
+We left off at theorizing how GPUs could potentially help accelerate these
+workflows.  Starting with Scikit-Image's implementation, we naively tried
+replacing `scipy.signal.convolve` calls with `cupyx.scipy.ndimage.convolve`,
+and while performance improved, it did not improve _significantly_ -- that is,
+we did not get the 100X speed we were looking for.
+
+As it often turns out in mathematics a problem that can be inefficient to solve
+in one representation can often be made more efficent by transforming the data
+beforehand. In this new representation we can solve the same problem
+(convolution in this case) more easily before transforming the result back into
+a more familiar representation. When it comes to convolution, the
+transformation we apply is called [Fast-Fourirer Transform
+(FFT)](https://en.wikipedia.org/wiki/Fast_Fourier_transform). Once this is
+applied we are able to convolve data using a simple multiplication.
+
+As it turns out this FFT transformation is extremely fast on both CPUs and
+GPUs. Similarly the algorithm we can write with FFTs are accelerated.  Despite
```

Comment:
The algorithm ... *is* accelerated.

Note though that to people "trained in the art" this whole paragraph may look a bit odd, as it is standard practice to do deconvolution in FFT space.

Reply:
One data point worth looking at here is SciPy actually chooses between doing it in real or FFT space ( https://github.com/scipy/scipy/pull/5608 ), which is [what scikit-image calls to do `richardson_lucy`]( https://github.com/scikit-image/scikit-image/blob/ed889ba3deb17cd94d1f0f9b730f6564585936ee/skimage/restoration/deconvolution.py#L379 ). So there may still be cases where real is more efficient.

Reply:
> Yeah that's fair. Would you recommend dropping it then?

I would suggest rephrasing it. Currently it reads a bit as "we discovered that one can actually do this in Fourier space and it turns out to be much faster"  to paraphrase it very loosely. I would simply phrase it more as "In contrast, to our previous example, we perform deconvolution in fourier space, a commonly used approach ...." and then keep some of the explanations why it is fast as they are still useful for readers who may not be familiar with the concept that a convolution of two signals in one domain is equivalent to multiplication of those signals in the Fourier-transformed domain.

Reply:
Ok I tried to add a line indicating this is a commonly used technique. Though this is also alluded to in the final sentence of the paragraph as well.

Reply:
Yes, I like that better.

Reply:
Thanks for taking a look :)